# August 2015 
# Modified September 2016
# Author: FBB

# reproduce the result in
http://www.mdrc.org/sites/default/files/What%20Strategies%20Work%20for%20the%20Hard%20FR.pdf 


In [6]:
import os
import sys
import numpy as np
import pylab as pl


import json
import os
#s = json.load( open(os.getenv("PUI2015")+"/fbb_matplotlibrc.json") )
#pl.rcParams.update(s)

%pylab inline

#from IPython.display import Image
#Image(filename='../plotsforclasses/NYCReentryprogram_title.png')



Populating the interactive namespace from numpy and matplotlib


In [7]:
#Image(filename='../plotsforclasses/NYCReentryprogram.png')

# NULL HYPOTHESIS: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05，$P_0$: Control group; $P_1$: Program group

# $H_0: P_0 - P_1 \geq$    0
    
# $H_a: P_0 - P_1 $< 0    
    
    
# $\alpha$ = 0.05    

## this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table above):
# $P_0=0.035, P_1=0.701$

In [8]:

alpha=0.05
# we like fractions better then percentages. as a rule of thumb, either use fractions or counts
P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0-P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

n_0 = 409
n_1 = 568

#lets get the counts by multiplying by the sample size
Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1


we must assess the statistical significance


# 2 samples, categorical data
# TWO OPTIONS z test, or chi-square test.  

# START WITH Z TEST

## the z test compares the stanrard deviation of the expected distribution and the observed result. it tells you literally how many standard deviations from the tail an observation is, under the _assumption of normality
must define the sample standard deviation 

In [9]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.422193449335


# standard deviation of the sampling distribution the distribution is Binomial, the binomail stdev is 

(see a proof here!: http://stats.stackexchange.com/questions/29641/standard-error-for-the-mean-of-a-sample-of-binomial-random-variables!): 

$\sqrt{\frac{p(1 - p)}{n}}$

for 2 samples this becomes 

$\sqrt{ \frac{ \hat{p}(1 - \hat{p})} {n0} + \frac{ \hat{p}(1 - \hat{p})} {n1} }$

cfr: page 138 of Statistics in a Nutshell, eq. 5.12 and here http://stattrek.com/hypothesis-test/difference-in-proportions.aspx?Tutorial=AP

### $\hat{p} =\frac{p_0  n_0 + p_1  n_1}{n_0+n_1}$


In [10]:
#Image("../plotsforclasses/SiNS5.13.png")

Note that in the Image above, taken from the online version of the book, $\bar{x}$  should be  $\hat{p}$!!

In [11]:

# I am goonna create a little one line function to calculate the standard error
# and to calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1)) 

# z score: how many standard deviation away from the population parameter is my statistic?

# $z=\frac{P_1-P_0}{\sigma}$

In [12]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)


20.7929348726


note that using p0-p1 or p1-p0 at the numerator is equivalent because the standardizes normal value of z has mean 0 (see image below) so that we can use the absolute value of the z score, or equivalently look for $P[Z<z]$ if z is positive, and $P[Z>z]$ if z is negative.

In [13]:
#Image('http://intersci.ss.uci.edu/wiki/images/3/3a/Normal01.jpg')

## if $p<\alpha$ : reject H0

## IMPORTANT!! note that this P in the bottom line of the table is not the p-value, but 
## p-value = 1-P

In [14]:
## p-value for employment after 2 years: 
## since the largest number we read off the table for is (way) smaller 
## than the value for our statistic 
## our p-value will be smaller than it would be if calculated using 
## (e.g.) .9998 (and in fact using 1.0000 which is the largest number 
## in the table). Using 0.9998 is a **conservative** approach. 

p_2y = 1 - 0.9998


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected


# !!!!!!!!TODO FOR YOU
## what if we used the values for where the former inmate was or was not "Convicted of a felony" (row 10) in the Recidivism (Years 1-3)?

# Null hypothesis: The probability of former prisoners covictied into a felony after release is the higher or the same for candidates who participated in the program as for the control group, significance level p = 0.05，$P_0$: Control group $P_1$: Program group 
# $H_0 = P_1 - P_0 \geq 0$
# $H_a = P_1 - P_0 < 0$ 

# $P_0 = 0.117, P_1= 0.1$

# look up data table and insert the appropriate values to get the appropriate result! you can use the functions I defined above, with different arguments. 

In [37]:
P_0 = 11.7/100
P_1 = 10.0/100
N_0 = 409.0
N_1 = 568.0


z_3y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_3y)




-0.84764020522


In [38]:
p_3y = 1 - 0.7995

report_result( p_3y , alpha)

is the p value 0.20 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected


# Now lets do it with the $\chi^2$ test

## this analysis can also be done with the $\chi^2$ test, and the  $\chi^2$ distribution, 
## see  [A. Marengo](http://www.csun.edu/~amarenco/Fcs%20682/When%20to%20use%20what%20test.pdf) on how to choose a test and "Statistics In a Nutshell Chapter 4", or http://math.hws.edu/javamath/ryan/ChiSquare.html (if you are really just interested in the formula at face value)

### The chisq statistics tests the statistics calculated as :

$\chi^2 = \sum_{i} \frac{(observation_i - expectation_i)^2}{expectation_i}$

### against a chi sq distribution.
### If we talk about sample fractions  that is 

$\chi^2 = \sum_i \frac{(f_{i,observed} - f_{i,expectated})^2}{f_{i,expected}}$

### Where _i_ indicates the sum over _each cell_.
### turns out this quantity is distributed according to a chi square distribution, so if i get the $\chi^2$ statistics i can compare it to the full chisq distribution and see how far in the tail it is

### The trickiest part (but not that tricky) is to figure out how to construct the table of values. please see Statistics In a Nutshell Chapter 4, for our data for example: Thisis called a CONTINGENCY TABLE

|                 |     success         | failure|    |               
|-----------------|:-------------------:|:-------------------:|---------------------------|
| test sample     | number of successes in test    | number of failures in test    | number members of test sample |
| control sample  | number of successes in control | number of failures in control | number members of control sample| 
|                 | total successes                |  total failures               | number of all members         |

|convicted in a felony      |     employed         | not employed    |                   
|---------------------------|:---------------------:|------------------|---------------------------|
| test sample               | $0.701*568$          | $0.299*568$       | 568                       |
| control sample            | $0.035*409$           | $0.965*409$      | 409                       |
|                           |                       |                  |                           |
| total                     | 412.483               |  564.517        | 977                       |

# for each cell the expected ratio is the product of the total of all rows and all columns, devided by the total

# FOR EACH CELL: 
# expected_cell = $\frac{row~tot * col~tot}{total}$

In [22]:
from evalChisq import evalChisq

In [43]:

Ntot = 977 # a + b + c + d = tot

sample_values = np.array([[0.701 * 568, 0.299 * 568], [0.0305 * 409, 0.965 * 409]])
chi_p1 = evalChisq(sample_values)
print (chi_p1)

437.186344001


# This number must be compared to the chi sq distribution. 
# You must calculate the number of degrees of freedom forthis experiment. 
# Generally: DOF = Number of observations - number of Independent Variables
# so here DOF = 1. Now you can look at the table below and draw conclusions about the rejection of the Null

In [44]:
#Image("http://passel.unl.edu/Image/Namuth-CovertDeana956176274/chi-sqaure%20distribution%20table.PNG")

## From the Chi square distribution table, we can find the number 3.84 for 0.05 siginificance level and 1 DOF.

In [45]:

def report_result_chi(p):
    a = 3.84
    print ('is the probability ' + 
           '{0:.2f} larger than the critical value {1:.2f}?'.format(p,a))
    if p > a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p > a  else 'not rejected') )
    


In [46]:
report_result_chi(chi_p1)

is the probability 437.19 larger than the critical value 3.84?
YES!
the Null hypothesis is rejected


437 is hella larger then 3.84

why am i mentioning 3.84?

how does the chi square statistic that you derived compare?

please  state what that means in terms of your Null hypothesis in a markdown cell below!

# !!!!!! TODO FOR YOU!
# lets see what the chi sq statistics says about the conviction for fellonies (row 10)

## fill in the contingency table below

|convicted of a fellony     |     yes   | no        |                |                 
|---------------------------|-----------|-----------|----------------|
| test sample               | $0.100*568$  | $0.9*568$        | 568                       |
| control sample            | $0.117*409$  | $0.883*409$      | 409                       |
|                           |              |                  |                           |
| total                     | 104.653      |  872.347         | 977                       |

In [47]:
Ntot = 977 # a + b + c + d = tot

sample_values = np.array([[0.1 * 568, 0.9 * 568], [0.117 * 409, 0.883 * 409]])
chi_p2 = evalChisq(sample_values)
print (chi_p2)

0.718493917505


In [48]:
report_result_chi(chi_p2)

is the probability 0.72 larger than the critical value 3.84?
NO!
the Null hypothesis is not rejected


## Which means that the NULL HYPOTHESIS: The probability of former prisoners covictied into a felony after release is the higher or the same for candidates who participated in the program as for the control group, can not be rejected at 0.05 significance level.